cpp/src/arrow/table.cc

```diff
@@ -415,6 +434,17 @@ Status Table::AddColumn(int i, const std::shared_ptr<Column>& col,
   return Status::OK();
 }
 
+Status Table::CastColumn(int i, const std::shared_ptr<DataType>& type,
+                         std::shared_ptr<Table>* out) const {
+  std::shared_ptr<Column> old_col = columns_[i];
+  std::shared_ptr<Column> new_col;
+  std::shared_ptr<Schema> new_schema;
+  RETURN_NOT_OK(old_col->cast(type, &new_col));
+  RETURN_NOT_OK(schema_->ReplaceField(i, new_col->field(), &new_schema));
+  *out = std::make_shared<Table>(new_schema, ReplaceVectorElement(columns_, i, new_col));
+  return Status::OK();
+}
```

Comment:
Similar comment to above. I would prefer

```
std::shared_ptr<Column> dictionary_column;
RETURN_NOT_OK(EncodeDictionary(table->column(i), memory_pool, &dictionary_column));
```

Since we may have many string columns in a table, it would be less expensive to build a new vector of columns, then create a new Table at the end